Fix bug 1184796: Optimize bibi mixin for multiple declarations

stephaniehobson opened the pull request:
**Current behaviour:**
```
.test {
    @include bidi-value(float, left, right);
    @include bidi-style(margin-right, ($grid-spacing / 2), margin-left, 0)
}
```
Outputs:
```
.test {
    float: left;
    margin-right: 10px;
}
html[dir='rtl'] .test {
    float: right;
}
html[dir='rtl'] .test {
    margin-right: 0;
    margin-left: 10px;
}
```

**Proposed behaviour:**
```
.test {
    @include bidi((
        (float, left, right),
        (margin-right, ($grid-spacing / 2), margin-left, 0)
    ));
}
```
Outputs:
```
.test {
    float: left;
    margin-right: 10px;
}
html[dir='rtl'] .test {
    float: right;
    margin-right: 0;
    margin-left: 10px;
}
```

This saves about 0.5KB in mdn.css after gzipping.

Questions:
- Is the new syntax confusing? Do we make contributions harder by changing?
- Is 0.5KB enough savings to keep working on this?

schalkneethling commented:
> Is the new syntax confusing? Do we make contributions harder by changing?

Maybe slightly but, if there is good documentation then I would say it is fair enough. I really like the cleaner grouped CSS that is output.

> Is 0.5KB enough savings to keep working on this?

I would say so. And if a small optimization such as this can yield such an impact, then I bet there are a few of these around the code base which will add up.

schalkneethling commented on kuma/static/styles/components/wiki/promos.scss:
```diff
@@ -150,8 +160,10 @@
-        @include bidi-style(padding, 20px 10px 140px, padding, 20px 10px 140px);
+        @include bidi((
+            (background-position, bottom 20px center, background-position, bottom 20px center),
+            (padding, 20px 10px 140px, padding, 20px 10px 140px)
```
Nit: If both the LTR and RTL style rules are the same, is there a specific reason we should specify it inside `bidi`?

stephaniehobson replied:
Yes, this is typically done when over-riding a value declared in a bidi further up the cascade, I will verify this and add notes in all the places this occurs :)

schalkneethling commented on kuma/static/styles/components/wiki/promos.scss:
```diff
@@ -168,8 +180,10 @@
-        @include bidi-style(padding, 20px 20px 160px, padding, 20px 20px 160px);
+        @include bidi((
+            (background-position, bottom center, background-position, bottom center),
+            (padding, 20px 20px 160px, padding, 20px 20px 160px)
```
Nit: Here as well, same question as above.

schalkneethling commented on kuma/static/styles/components/wiki/promos.scss:
```diff
@@ -201,8 +215,10 @@
-        @include bidi-style(padding, 0 0 175px, padding, 0 0 175px);
+        @include bidi((
+            (background-position, bottom center, background-position, bottom center),
+            (padding, 0 0 175px, padding, 0 0 175px)
```
Nit: Also here. This is the third instance so far so, I am starting to think there is a reason why it is done this way?

schalkneethling commented on kuma/static/styles/includes-skinny/_mixins.scss:
```diff
@@ -332,6 +332,27 @@ Spacing
 Bidi / l10n
 ====================================================================== */
 
+@mixin bidi($list) {
```
Would be great to add some documentation here with a sample of a use case and the code that is produced.

stephaniehobson commented:
I replaced all occurrences of bidi-style and bidi-value with this mixin and removed those mixins. Now we do it the same way everywhere. I hope that will make it easier for new contributors.

I found a problem where single line declarations don't always work right if they're not properly formatted and proper formatting is not intuitive:
`@include bidi(((float, right, left),));`

All occurrences are correct now though so if someone copies and pastes they should be able to just change the values.

Ready for second santiy-check and review.

schalkneethling commented on kuma/static/styles/includes/_mixins.scss:
```diff
@@ -330,26 +332,108 @@ Spacing
+                /*
+                    Trouble shooting:
+                    - Check that you have the right number of nested (((parentheses)))
+                    - Check that there is a trailing comma
```
@stephaniehobson Should there always be a trailing comma? Or only in these instances:

```
@include bidi(((float, left, right),));
```

stephaniehobson replied:
Yes, there should always be a trailing comma.

Technically, it works in some situations without it but we should be consistent.

stephaniehobson replied:
In case you're curious: it seems to work without the comma if there are no spaces or parentheses in the declaration.

schalkneethling commented on kuma/static/styles/components/wiki/promos.scss:
```diff
@@ -41,19 +43,21 @@
     @include clearfix();
-    @include bidi-style(background-position, bottom 20px right 20px, background-position, bottom 20px left 20px);
-    @include bidi-style(padding, 20px 350px 20px 25px, padding, 20px 25px 20px 350px);
+     @include bidi((
```
Nit: indentation fails stylelint

stephaniehobson commented:
Nit fixed, squashed, double checked for trailing commas, rebased. Ready to be merged when reviews are passed!
